samtranslator/model/sam_resources.py

```diff
@@ -802,6 +803,126 @@ def _validate_deployment_preference_and_add_update_policy(
+        canary.SuccessRetentionPeriod = self.SuccessRetentionPeriod
+        canary.Tags = self._construct_tag_list(self.Tags)
+        canary.VPCConfig = self.VpcConfig
+
+        if self.Tracing or self.Environment or self.MemorySize or self.Timeout:
+            canary.RunConfig = self._construct_run_config()
+        return canary
+
+    def _extract_not_none_properties(self, dict):
+        """
+        Filters out not None properties
+        """
+        return {k: v for k, v in dict if v is not None}
+
+    def _construct_run_config(self):
+        """
+        If the user specifies any of Tracing, MemorySize, TImeout or Environement then the RunConfig resource in the
+        transformed AWS::Synthetics::Canary needs to be added. Note, for Environment property the syntaxt in AWS::Serverless::Canary is
+        Environment:
+            Variables:
+                Var1: Var2
+        while in AWS::Synthetics::Canary its
+        EnvironmentVariables:
+            Var1: Var2
+        so it needs to be transformed accordingly
+        """
+        runconfig = {"ActiveTracing": self.Tracing, "MemoryInMB": self.MemorySize, "TimeoutInSeconds": self.Timeout}
+        if self.Environment:
+            runconfig["EnvironmentVariables"] = self.Environment["Variables"]
+
+        return self._extract_not_none_properties(runconfig.items())
+
+    @property
+    def _construct_code_dict(self):
+        """Constructs Synthetics Canary Code Dictionary based on the accepted SAM artifact properties such
+        as `InlineCode` and `CodeUri`
+
+        :raises InvalidResourceException when conditions on the SAM artifact properties are not met.
+        """
+        # accepted artifacts
+        artifacts = {"InlineCode": self.InlineCode, "CodeUri": self.CodeUri}
+
+        # Inline function for transformation of inline code.
+        # It accepts arbitrary argumemnts, because the arguments do not matter for the result.
+        def _construct_inline_code(*args, **kwargs):
+            return {"Handler": self.Handler, "Script": self.InlineCode}
+
+        # dispatch mechanism per artifact on how it needs to be transformed.
+        artifact_dispatch = {
+            "InlineCode": _construct_inline_code,
+            "CodeUri": construct_s3_location_object,
+        }
+
+        filtered_artifacts = self._extract_not_none_properties(artifacts.items())
+        filtered_artifact_keys = list(filtered_artifacts.keys())
+
+        if "InlineCode" in filtered_artifact_keys:
+            set_artifact = "InlineCode"
+        elif "CodeUri" in filtered_artifact_keys:
+            set_artifact = "CodeUri"
+        else:
+            raise InvalidResourceException(self.logical_id, "Either 'InlineCode' or 'CodeUri' must be set.")
+
+        dispatch_function = artifact_dispatch[set_artifact]
+        code = dispatch_function(artifacts[set_artifact], self.logical_id, set_artifact)
+        code["Handler"] = self.Handler
+        return code
```

Comment:
Seems these logics can be rearranged to make it clearer:

```suggestion
        filtered_artifacts = self._extract_not_none_properties(artifacts.items())
        filtered_artifact_keys = list(filtered_artifacts.keys())

        if "InlineCode" in filtered_artifact_keys:
            return {"Handler": self.Handler, "Script": self.InlineCode}
        elif "CodeUri" in filtered_artifact_keys:
            code = construct_s3_location_object(self.CodeUri, self.logical_id, "CodeUri")
            code["Handler"] = self.Handler
            return code
        else:
            raise InvalidResourceException(self.logical_id, "Either 'InlineCode' or 'CodeUri' must be set.")
```